discovery/src/main/java/com/alibaba/fescar/discovery/registry/consul/ConsulRegistryServiceImpl.java

```diff
@@ -0,0 +1,336 @@
+        newService.setTags(Collections.singletonList(SERVICE_TAG));
+        newService.setPort(address.getPort());
+        newService.setAddress(NetUtil.toIpAddress(address));
+        newService.setCheck(createCheck(address));
+        return newService;
+    }
+
+    /**
+     * create service check based on TCP
+     *
+     * @param address
+     * @return
+     */
+    private NewService.Check createCheck(InetSocketAddress address) {
+        NewService.Check check = new NewService.Check();
+        check.setTcp(NetUtil.toStringAddress(address));
+        check.setInterval(DEFAULT_CHECK_INTERVAL);
+        check.setTimeout(DEFAULT_CHECK_TIMEOUT);
+        check.setDeregisterCriticalServiceAfter(DEFAULT_DEREGISTER_TIME);
+        return check;
+    }
+
+    /**
+     * get healthy services
+     *
+     * @param service
+     * @return
+     */
+    private Response<List<HealthService>> getHealthyServices(String service, long index, long watchTimeout) {
+        return getConsulClient().getHealthServices(service, HealthServicesRequest.newBuilder()
+            .setTag(SERVICE_TAG)
+            .setQueryParams(new QueryParams(watchTimeout, index))
+            .setPassing(true)
+            .build());
+    }
+
+    /**
+     * get service group
+     *
+     * @param key
+     * @return clusterNameKey
+     */
+    private String getServiceGroup(String key) {
+        String clusterNameKey = PREFIX_SERVICE_ROOT + CONFIG_SPLIT_CHAR + PREFIX_SERVICE_MAPPING + key;
+        return FILE_CONFIG.getConfig(clusterNameKey);
+    }
+
+    /**
+     * refresh cluster
+     *
+     * @param cluster
+     */
+    private void refreshCluster(String cluster) {
+        if (null == cluster) {
+            return;
+        }
+        Response<List<HealthService>> response = getHealthyServices(getClusterName(), -1, -1);
+        if (response == null) {
+            return;
+        }
+        refreshCluster(cluster, response.getValue());
+    }
+
+    /**
+     * refresh cluster
+     *
+     * @param cluster
+     * @param services
+     */
+    private void refreshCluster(String cluster, List<HealthService> services) {
+        if (null == cluster || services == null) {
+            return;
+        }
+        clusterAddressMap.put(cluster, services.stream()
+            .map(HealthService::getService)
+            .map(service -> new InetSocketAddress(service.getAddress(), service.getPort()))
+            .collect(Collectors.toList()));
+    }
+
+    /**
+     * consul notifier
+     */
+    private class ConsulNotifier implements Runnable {
+        private String cluster;
+        private long consulIndex;
+        private boolean running;
+
+        ConsulNotifier(String cluster, long consulIndex) {
+            this.cluster = cluster;
+            this.consulIndex = consulIndex;
+            this.running = true;
+        }
+
+        @Override
+        public void run() {
+            while (this.running) {
```

Comment:
it seems the stop may never invoke ？so this running is always true

Reply:
it will stop when client exits or the unsubscribe method is invoked.

Reply:
ok